provider/oci/environ.go

```diff
@@ -223,42 +378,313 @@ func (e *Environ) Destroy(ctx envcontext.ProviderCallContext) error {
+		ids[i] = val.Id()
+	}
+
+	err = e.StopInstances(ctx, ids...)
+	if err != nil {
+		return err
+	}
+	logger.Debugf("Cleaning up network resources")
+	return e.cleanupNetworksAndSubnets(controllerUUID, "")
 }
 
 // Provider implements environs.Environ.
 func (e *Environ) Provider() environs.EnvironProvider {
-	return nil
+	return e.p
 }
 
-// StorageProviderTypes implements storage.ProviderRegistry.
-func (e *Environ) StorageProviderTypes() ([]storage.ProviderType, error) {
-	return []storage.ProviderType{}, errors.NotImplementedf("StorageProviderTypes")
-}
+// getCloudInitConfig returns a CloudConfig instance. The default oracle images come
+// bundled with iptables-persistent on Ubuntu and firewalld on CentOS, which maintains
+// a number of iptables firewall rules. We need to at least allow the juju API port for state
+// machines. SSH port is allowed by default on linux images.
+func (e *Environ) getCloudInitConfig(series string, apiPort int) (cloudinit.CloudConfig, error) {
+	// TODO (gsamfira): remove this function when the above mention bug is fixed
+	cloudcfg, err := cloudinit.New(series)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot create cloudinit template")
+	}
 
-// StorageProvider implements storage.ProviderRegistry.
-func (e *Environ) StorageProvider(t storage.ProviderType) (storage.Provider, error) {
-	return nil, errors.NotFoundf("storage provider %q", t)
+	if apiPort == 0 {
+		return cloudcfg, nil
+	}
+
+	operatingSystem, err := jujuseries.GetOSFromSeries(series)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	switch operatingSystem {
+	case os.Ubuntu:
+		fwCmd := fmt.Sprintf(
+			"/sbin/iptables -I INPUT -p tcp --dport %d -j ACCEPT", apiPort)
+		cloudcfg.AddRunCmd(fwCmd)
+		cloudcfg.AddScripts("/etc/init.d/netfilter-persistent save")
+	case os.CentOS:
+		fwCmd := fmt.Sprintf("firewall-cmd --zone=public --add-port=%d/tcp --permanent", apiPort)
+		cloudcfg.AddRunCmd(fwCmd)
+	}
+	return cloudcfg, nil
 }
 
 // StartInstance implements environs.InstanceBroker.
 func (e *Environ) StartInstance(ctx envcontext.ProviderCallContext, args environs.StartInstanceParams) (*environs.StartInstanceResult, error) {
-	return nil, errors.NotImplementedf("StartInstance")
+	// var types []instances.InstanceType
+
+	if args.ControllerUUID == "" {
+		return nil, errors.NotFoundf("Controller UUID")
+	}
+
+	networks, err := e.ensureNetworksAndSubnets(ctx, args.ControllerUUID, e.Config().UUID())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	zones, err := e.AvailabilityZones(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	zone := zones[0].Name()
+	network := networks[zone][0]
+	// refresh the global image cache
+	// this only hits the API every 30 minutes, otherwise just retrieves
+	// from cache
+	imgCache, err := refreshImageCache(e.Compute, e.ecfg().compartmentID())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	logger.Debugf("Image cache contains: %v", imgCache)
```

Comment:
Have you bootstrapped with "--debug" to see if this is useful? It feels like we might prefer this to be:
logger.Tracef("Image cache contains: %# v", pretty.Formatter(imgCache))

that way we get better rendering when we do need it, but Debug seems a bit too verbose.